Differentiate Floodgate players in database

Smart123s opened the pull request:
### Summary of your change
Please read #700.
Added a new column to the database called `Floodgate`.
This will make the life of Bedrock and Java players with the same name a lot easier.

### Why Draft
I've just started working on this, and I wanted to share the code while I'm progressing with it.
***<ins>At this stage you have to delete the old database, or the server will crash when someone tries to join it.</ins>***

### Todo
- [x] Database migrator (so the old database won't have to be deleted)
- [x] Migrate every part of the code to be aware of the new `Floodgate` field in the database
  - [x] ~~Migrate `activate` plugin message~~*
  - [x] ~~Migrate `/cracked` and `/premium` commands (self)~~*
    - [x] Fix issue described at "Floodgate detection" (issue #689)
  - [x] ~~Migrate `/cracked` and `/premium` commands (others)~~*
- [x] Check what the `Premium` field in the db does for Floodgate players (so check for leftover code)
- [ ] Debug why `uuid` and `lastIp` are `null` for Floodgate players in the database ([code](https://github.com/games647/FastLogin/blob/8a01ddc231d6981be4ed755d3669727fc3bed4f5/core/src/main/java/com/github/games647/fastlogin/core/shared/FloodgateManagement.java#L171-L172))

_*No longer needed, becuase usernames are now unique (https://github.com/games647/FastLogin/pull/709#issuecomment-1184623441)._

### To Test
*just notes for myself, so that I won't forget it*
- [x] Server without Floodgate
- [x] BungeeCord
- [x] Premium/Cracked commands
- [x] [AsyncToggleMessage](https://github.com/games647/FastLogin/blob/9656aadb353b0061b40f405cecab3481cd422738/bungee/src/main/java/com/github/games647/fastlogin/bungee/task/AsyncToggleMessage.java)
- [x] Auto login/register with same name Java and Bedrock accounts
    Result: only works on the platform where premium login was enabled

### Floodgate detection
https://github.com/games647/FastLogin/issues/689 needs to be fixed. (I'm not linking that issue to his PR on purpose, as it's not fixed yet.)
In short, if ProtocolLib prefix workaround is enabled, the following code wont work, after the prefix is added:
https://github.com/games647/FastLogin/blob/9b04ea5c89e0b99bc273d28a1d9b0483c0a1b903/core/src/main/java/com/github/games647/fastlogin/core/hooks/bedrock/FloodgateService.java#L105-L114

### Related issue
Fixes https://github.com/games647/FastLogin/issues/700 
Fixes https://github.com/games647/FastLogin/issues/696
Fixes https://github.com/games647/FastLogin/issues/633

Smart123s commented:
@games647 I've got some questions.

As of now, (on branch `main`/https://github.com/games647/FastLogin/commit/9b04ea5c89e0b99bc273d28a1d9b0483c0a1b903), FastLogin allows Floodgate and Java players to be registered with the same name (without prefixes). However, the database contains

https://github.com/games647/FastLogin/blob/af2df500577db8a01c9cf664477fa11921eb5e97/core/src/main/java/com/github/games647/fastlogin/core/storage/SQLStorage.java#L59-L60

which forbids registering two players with the same name.

While I am strongly **against** registering multiple players with the same name, because it can cause confusion with some plugins (especially auth plugins), but there are folks out there already using this approach. For example, see https://github.com/games647/FastLogin/issues/696.

I have came up with the following possibilities as a solution:
- Deny joining the server, if the name has been registered (=saved in db) from a different platform (=Floodgate/Java)
  - This would be a bit forcing in my opinion, considering that Floodgate *allows* that (via disabling username prefixes)
- Allow auto login/registering only for the platform the player (name) joined first
  - With the approach, the player from second platform doesn't have to be saved in the db
- Remove the `UNIQUE` constraint from the db table, and allow same named registrations from multiple platforms (as it was before)
  - Are there any checks that should be introduced in other parts of the code, if the `UNIQUE` constraint is removed?

games647 commented:
To be honest name changing is difficult topic. I made the `UNIQUE` decision earlier, because we try to find the player profile on login based on the player name. If there are two, who should we pick? Allowing name changes would mean the old name is longer up to date and should be removed, therefore unique.

Correct me if I'm wrong, but the referenced ticket is about the login with the same player name, that is the same account owner? The best option would be to allow both, because Floodgate authentication is identical to onlinemode verification. It only differs that FastLogin doesn't perform the validation. However, I noticed, this should only apply to linked profiles. Otherwise it should be forbidden for the second platform. 

Although it would be possible to have two accounts with the same player name, because as far as I know Floodgate uses different UUID. Nevertheless it could cause conflicts like you mentioned.

Smart123s commented:
Note for myslef: https://github.com/games647/FastLogin/pull/709/commits/87b21fd7d01f22062cf24c9887b53e045c327c61 (and it's parents) have merge conflicts resolved with https://github.com/games647/FastLogin/commit/6beaf194ce93ef0ae21f7e71286717f15ab54c18. This branch is currently based on https://github.com/games647/FastLogin/commit/62d7320a6e3f37c6b1a0ef63456c29856e80c9d6, as I couldn't strat up a server with anything newer.

Smart123s commented:
This pull request is more or less ready for review. I'll finish the remaining TODOs, run some test, and probably do a rebase.

Commits 1c63e70, 25e71e8, ae9772c, 24d552d, eab6ef9 could have been a separate pull request (I can still open a new one, if you'd like me to), but I thought that they fix an important issue (the floodgate prefix bug with ProtocolLib) that it should be fixed before the new database columns get added.

I might replace the [getBedrockPlayer()](https://github.com/games647/FastLogin/blob/40ac3733944d845cab2f07d473820c810ff2a4db/core/src/main/java/com/github/games647/fastlogin/core/hooks/bedrock/FloodgateService.java#L98-L114) function with the channel extraction method from https://github.com/games647/FastLogin/pull/709/commits/1c63e7037c219ac1beee5ae6e8715f01e366f313 and https://github.com/games647/FastLogin/pull/709/commits/ae9772c0b028c47367961eb000f3ab50e39cf940, but that could be done in a future PR. It'd also require me to figure out how to extract the player's channel in Bungee and in ProtocolSupport (and Velocity in the future), which would take some time.

For migrating legacy databases I came up with the idea to add a new integer column called `DBVersion`. If the database will have to be updated again in the future, we could simply increase the value of DBVersion. Is this approach OK?

<hr>

Do you have any idea on how to make this line shorter?

https://github.com/games647/FastLogin/blob/ae9772c0b028c47367961eb000f3ab50e39cf940/bukkit/src/main/java/com/github/games647/fastlogin/bukkit/listener/protocollib/ProtocolLibListener.java#L295

This didn't help: https://stackoverflow.com/questions/41056517/breaking-a-long-url-to-several-line-in-javadoc

games647 commented:
>For migrating legacy databases I came up with the idea to add a new integer column called DBVersion. If the database will have to be updated again in the future, we could simply increase the value of DBVersion. Is this approach OK?

Well there also the approach of doing [migrations](https://book.cakephp.org/migrations/3/en/index.html). This what I heard is normally used when doing schema changes. However I don't know if they fit here, because we are doing a decision during the login process and not on the data itself.

>Do you have any idea on how to make this line shorter?

Well you can split the `a`-tag, but that's it. 
```xml
<a href="">
lorem ipsum
</a>
```

Attribute `href` could also be on the next line, but that just looks strange.

Smart123s commented:
> Well there also the approach of doing [migrations](https://book.cakephp.org/migrations/3/en/index.html). This what I heard is normally used when doing schema changes.

The pluign you provided uses PHP, and I don't think that it'd fit for a java project. However, [Flyway](https://flywaydb.org/documentation/tutorials/java) could be used, which has a Java API.

> However I don't know if they fit here, because we are doing a decision during the login process and not on the data itself.

There are three approaches I could think of:

1. (current) Have a DBVersion column to check if the player was migrated or not.
 - tbh, I'm starting to think that this is just unnecessary bloat
2. Set Floodgate as a nullable boolean, and use that instead of a DBVersion column
 - and maybe make it nullable if a new database is created (so it's not an old one being migrated), although that would fragment the db structure, which could cause headaches in the future
3. Keep Floodgate as `not null` and tell everyone who updates the plugin that Floodgate players have to log in using their passwords, and type `/premium` to migrate themselves.
 - this would be the best idea in the long term
 - Floodgate support is still experimental, it's written 5 times in the config ([1](https://github.com/games647/FastLogin/blob/833177933a2e670b6ef6a49ee0c44d426672f432/core/src/main/resources/config.yml#L214-L215), [2](https://github.com/games647/FastLogin/blob/833177933a2e670b6ef6a49ee0c44d426672f432/core/src/main/resources/config.yml#L226-L227), [3](https://github.com/games647/FastLogin/blob/833177933a2e670b6ef6a49ee0c44d426672f432/core/src/main/resources/config.yml#L249-L250), [4](https://github.com/games647/FastLogin/blob/833177933a2e670b6ef6a49ee0c44d426672f432/core/src/main/resources/config.yml#L263-L264), [5](https://github.com/games647/FastLogin/blob/833177933a2e670b6ef6a49ee0c44d426672f432/core/src/main/resources/config.yml#L272)), so I could use it as an excuse to tell people that I don't care if a Floodgate player hasn't written down their automatically registered password. Although I'm pretty sure there would be a lot of new issues opened, and lots of people would kindly ask me to go to hell.

Which one should I go with? I'd actually prefer the second one. Regardless, I could use [Flyway](https://flywaydb.org/documentation/tutorials/java) to add the missing column(s) to the database. Would that plugin/library be okay?

games647 commented:
I rather was talking about the concept of having change based migrations. So one table which is used to keep track of the applied migrations and then seperate files for each migration. The implementation can be very simple, because the files can fully be read and then applied. So we then have something like:

migrations/01_base_table.sql (the current table for existing setups)
migrations/02_add_floodgate_differentiation.sql
migrations/03_track_name_changes.sql (example)
migrations/04_drop_something.sql (example)

I don't think we need a library for that. We check the number of applied migrations and then apply the sql file(s) of which are missing.

What do you think about that? This might be a bit overkill, because it's often used in combinations with many contributors (then also with timestamps) and frequent scheme changes.

>and maybe make it nullable if a new database is created (so it's not an old one being migrated), although that would fragment the db structure, which could cause headaches in the future

I guess that's the best option to migrate the old data while using the new structure for new players. We could later drop the backwards compatibility by migrating all nullable values to false.

games647 commented on bukkit/pom.xml:
```diff
@@ -376,7 +376,6 @@
             <groupId>io.netty</groupId>
             <artifactId>netty-transport</artifactId>
             <version>${nettyVersion}</version>
-            <scope>test</scope>
```
Could you check if this one gets shaded into the plugin? It might should be provided if we only reference on it.

Smart123s replied:
Moved to & fixed in #876

games647 reviewed (changes requested):
Looks good so far

Smart123s commented:
> I rather was talking about the concept of having change based migrations.

Then I misunderstood you, I thought you were talking about the library.

> I don't think we need a library for that.

Me neither.

> This might be a bit overkill

It'd make code cleaner, and it'd also make potential future migrations easier, so I think it's fine.

> I guess that's the best option to migrate the old data while using the new structure for new players. We could later drop the backwards compatibility by migrating all nullable values to false.

This will be the way then.

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/storage/SQLStorage.java:
```diff
@@ -90,6 +97,57 @@ public void createTables() throws SQLException {
+            try (PreparedStatement saveStmt = con.prepareStatement(floodgateStatement)) {
+                saveStmt.execute();
+            }
+            try (PreparedStatement saveStmt = con.prepareStatement(dbRevStatement)) {
```
You don't need a prepared statement here, because there are no placeholders.

games647 commented on bukkit/src/main/java/com/github/games647/fastlogin/bukkit/listener/protocollib/ProtocolLibListener.java:
```diff
@@ -254,4 +267,65 @@ private String getUsername(PacketContainer packet) {
+        AttributeKey<String> kickMessageAttribute = AttributeKey.valueOf("floodgate-kick-message");
+        String kickMessage = channel.attr(kickMessageAttribute).get();
+        if (kickMessage != null) {
+            packetEvent.getPlayer().kickPlayer(kickMessage);
```
Use the player variable from above directly.

Smart123s replied:
Moved to & fixed in #876

Smart123s commented:
I've moved the ProtocolLib fix to https://github.com/games647/FastLogin/pull/876 so I won't have to rebase ProtocolLib related commits when working on the database. As long as the fix gets merged before the database modifications, things should be fine.

Smart123s commented:
The migrator works, but the code is a mess in my opinion. 

I think that it'd be better to refactor the code, and make SQLStorage a kind of connection manager (initialize connection to the db) and make a separate class for every table. There classes would access `dataSource` from the common SQLStorage class. Should I go for it?

<hr>

Additionally, is the `ReentrantLock` in SQLIteStorage needed? I think that Hikari/JDBC makes sure that only one connection is open to the database on its own. By accident I forgot to close a `Connection` in MigrationManager after executing a query, and and upon the first login attempt (or the first call to  `dataSource.getConnection();`) I got the following error:

```
[16:59:59] [Craft Scheduler Thread - 6 - FastLogin/ERROR]: [FastLogin] Failed to query profile: .Smart123ss
java.sql.SQLTransientConnectionException: FastLogin - Connection is not available, request timed out after 30010ms.
	at fastlogin.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696) ~[FastLoginBukkit.jar:?]
	at fastlogin.hikari.pool.HikariPool.getConnection(HikariPool.java:197) ~[FastLoginBukkit.jar:?]
	at fastlogin.hikari.pool.HikariPool.getConnection(HikariPool.java:162) ~[FastLoginBukkit.jar:?]
	at fastlogin.hikari.HikariDataSource.getConnection(HikariDataSource.java:100) ~[FastLoginBukkit.jar:?]
	at com.github.games647.fastlogin.core.storage.SQLStorage.loadProfile(SQLStorage.java:107) ~[FastLoginBukkit.jar:?]
	at com.github.games647.fastlogin.core.storage.SQLiteStorage.loadProfile(SQLiteStorage.java:83) ~[FastLoginBukkit.jar:?]
	at com.github.games647.fastlogin.core.shared.FloodgateManagement.run(FloodgateManagement.java:83) ~[FastLoginBukkit.jar:?]
	at org.bukkit.craftbukkit.v1_19_R1.scheduler.CraftTask.run(CraftTask.java:101) ~[paper-1.19.1.jar:git-Paper-110]
	at org.bukkit.craftbukkit.v1_19_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.19.1.jar:git-Paper-110]
	at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.19.1.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
[16:59:59] [Craft Scheduler Thread - 6 - FastLogin/WARN]: [FastLogin] Plugin FastLogin v1.12-SNAPSHOT-a47b879 generated an exception while executing task 17
java.lang.NullPointerException: Cannot invoke "com.github.games647.fastlogin.core.StoredProfile.isSaved()" because "this.profile" is null
	at com.github.games647.fastlogin.core.shared.FloodgateManagement.run(FloodgateManagement.java:84) ~[FastLoginBukkit.jar:?]
	at org.bukkit.craftbukkit.v1_19_R1.scheduler.CraftTask.run(CraftTask.java:101) ~[paper-1.19.1.jar:git-Paper-110]
	at org.bukkit.craftbukkit.v1_19_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.19.1.jar:git-Paper-110]
	at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.19.1.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

*Note: I've rebased the code multiple times since I encountered this issue (and fixed it), so I can't provide the exact lines and context where the issue occurred. I didn't test MySQL for this issue.*

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/shared/JoinManagement.java:
```diff
@@ -138,8 +152,14 @@ private boolean checkPremiumName(S source, String username, StoredProfile profil
             if (storedProfile != null) {
+                if (storedProfile.isFloodgate()) {
+                    core.getPlugin().getLog()
+                            .warn("Player {} is already stored by FastLogin as a Bedrock Edition player.", username);
```
```suggestion
                            .info("Player {} is already stored by FastLogin as a Bedrock Edition player.", username);
```

Sounds more informative than warning the server owner.

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/shared/JoinManagement.java:
```diff
@@ -50,19 +50,33 @@ public JoinManagement(FastLoginCore<P, C, ?> core, AuthPlugin<P> authHook, Bedro
+            core.getPlugin().getLog().info(
+                    "Player {} will be migrated to the v2 database schema as a JAVA user", username);
+        } else if (profile.isFloodgate()) {
+            core.getPlugin().getLog().warn("Player {} is already stored by FastLogin as a Bedrock Edition player",
```
```suggestion
            core.getPlugin().getLog().info("Player {} is already stored by FastLogin as a Bedrock Edition player",
```

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/shared/FloodgateManagement.java:
```diff
@@ -76,11 +76,38 @@ public void run() {
+            // linked players are stored as Java (= not Floodgate) players
+            profile.setFloodgate(!isLinked);
+        } else {
+            if (!profile.isFloodgateMigrated()) {
```
Please put the non-inverted case first. It's easier to follow the flow.

Smart123s replied:
Resolved in 35c289d26d115ffaaafdc02dc09f36f1b02ce175

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/shared/FloodgateManagement.java:
```diff
@@ -76,11 +76,38 @@ public void run() {
+                            "Player {} will be migrated to the v2 database schema as a Floodgate user", username);
+                }
+            } else if (!profile.isFloodgate() && !isLinked) {
+                core.getPlugin().getLog().warn("Player {} is already stored by FastLogin as a Java Edition player",
```
```suggestion
                core.getPlugin().getLog().info("Player {} is already stored by FastLogin as a Java Edition player",
```

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/shared/FloodgateManagement.java:
```diff
@@ -76,11 +76,38 @@ public void run() {
+                return;
+            } else if (profile.isFloodgate() && isLinked) {
+                core.getPlugin().getLog()
+                        .warn("Player {} is already stored by FastLogin as a non-linked Bedrock Edition player",
```
```suggestion
                        .info("Player {} is already stored by FastLogin as a non-linked Bedrock Edition player",
```

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/shared/FloodgateManagement.java:
```diff
@@ -76,11 +76,38 @@ public void run() {
+        profile = core.getAuthStorage().loadProfile(username);
+        if (!profile.isSaved()) {
+            // linked players are stored as Java (= not Floodgate) players
+            profile.setFloodgate(!isLinked);
```
Why do you considered this approach? I feel like setting this to `false` could lead to the same issues. Is there any issue leaving it true and then read the configuration instead? This would make it easier to change configuration files.

Maybe we could make this field also a tristate.

Smart123s replied:
Linked Floodgate players are 99% Java players: they have the same name (without prefix), the same UUID, inventory, etc. The only difference is that they are present in the Floodgate API and they can't reply to premium java login requests.
I decided to keep it `false`, because they are the same player as their Java part, and I wanted to avoid storing one player as two entries in the database.

Smart123s replied:
I've added 5d994d8, but it's still a mess.

Let's say we have a player called Steve, who plays from Java, and linked Floodgate too. What should the value of Floodgate be in our database? Linked or Java? Both of them are correct. Names are unique in the database, so there's no way to have two row for one player.

games647 replied:
I think I'm lost. Why would we need two rows? You mean for the case Floodgate and Java profiles are associated to different accounts/persons? This would make it very difficult, because the auth plugin has only one single account often based on the player name. However, auth plugins with UUID support would in fact be different. 

The reason I thought name is unique, is because the only a single user can have this Mojang account name. If a user joins with the same name, but a different UUID, this would mean that the old data name is now invalid.

Smart123s replied:
I want to keep names unique, changing it would cause a lot of headaches.

What I tried to explain is that Java and their linked Floodgate accounts have the same name (and UUID), and I don't know how to save them in the database.

For example, I join a brand new server from my PC with my Java account. Then I get saved in the database as `Smart123s`, `FloodgateState.JAVA`. Then I get on the bus, grab my phone, and join the server with my linked Floodgate account. My name is still `Smart123s`, because I'm linked, but I'm no longer `FloodgateState.JAVA`, because I'm joining from bedrock this time. Should the database be changed to `Smart123s`, `FloodgateState.LINKED`? And then what happens if I join from Java again?

That's what I initially wanted to avoid by storing linked Floodgate players as Java players.

games647 replied:
>Should the database be changed to Smart123s, FloodgateState.LINKED? And then what happens if I join from Java again?

Well we could keep it as linked as it would mean (Floodgate+Java), so it doesn't represent the current session, but the account/person that is behind it.

Smart123s replied:
Ok, but there are two minor caveats. Firstly, it's impossible to check if a new Java player has a linked Floodgate account until he joins with it. And it's also impossible to detect if a Floodgate account was unlinked. Technically, we could send API requests to Geyser to check for these, but it'd slow down the login process without any actual benefit.

Anyways, it's gonna be a busy week/weekend for me, but I'll check on the code to make sure it works as we discussed above.

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/storage/MigratableStorage.java:
```diff
@@ -0,0 +1,43 @@
+     * @param currentVersion The current version of the table
+     * @return an SQL statement to migrate the table to the next version
+     */
+    String getMigrationStatement(int currentVersion);
```
Wouldn't it be better to have the migrations as files or separate Java classes? For example for the case a migration performs multiple statements.

Smart123s replied:
I have added https://github.com/games647/FastLogin/pull/709/commits/b59cc66d1091c3836d8119636260d7c30573250c, could you please check it?

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/storage/MigrationManager.java:
```diff
@@ -0,0 +1,124 @@
+public class MigrationManager {
+
+    protected static final String MIGRATION_TABLE = "migrations";
+    protected static final String CREATE_TABLE_STMT = "CREATE TABLE IF NOT EXISTS `" + MIGRATION_TABLE + "` ("
```
Migration approaches often use one row for each migration that you can individually rollback. I think this could be overkill for this project, but at least we should short comment what this approach does or explain the table scheme.

Smart123s replied:
Is 69df3a58d79c054c4f38c1bf4cb9692ca2986624 enough?

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/storage/MigrationManager.java:
```diff
@@ -0,0 +1,124 @@
+            initialVersion = 1;
+        }
+
+        for (int i = initialVersion; i < table.getLatestTableVersion(); i++) {
```
We should better differentiate between what `latest` or `getTableVersion` means. Maybe something like `getCurrentVersion` and `getRequiredVersion`. Or add a comment that makes it clear.

Smart123s replied:
Changed in e401103bc93a0d365499887672a99b46a3bcb752

games647 reviewed (changes requested):
Looks very good so far.

Smart123s commented on core/src/main/java/com/github/games647/fastlogin/core/storage/AuthStorage.java:
```diff
@@ -35,6 +35,4 @@ public interface AuthStorage {
 
     void save(StoredProfile playerProfile);
-
-    void close();
```
I moved it from AuthStorage to

https://github.com/games647/FastLogin/blob/06418b637ce27d31e1d4365a5d56d491add3142b/core/src/main/java/com/github/games647/fastlogin/core/storage/SQLStorage.java#L74-L76

games647 replied:
Well I think it's part of a functionality of the interface and not of the specific implementation of SQLStorage. To be precise, cases like FileStorage then should also have a close method.

Smart123s replied:
Is 3ffb9c7b18c8cb730078c75943cff407a615888b ok, or should I remove the close method from SQLStorage and reference `storage.dataSource` directly in SQLAuthStorage?

Smart123s commented on core/src/main/java/com/github/games647/fastlogin/core/storage/MigrationManager.java:
```diff
@@ -0,0 +1,124 @@
+        int initialVersion = getTableVersion(table.getTableName());
+
+        // special case: table premium was created before the migration manager
+        if (initialVersion == 0 && "premium".equals(table.getTableName()) && table.tableExists()) {
```
Resolved in 0dae497b9d594271acdf62a0fa767beb112869c9

Smart123s commented:
I have tested the migration from 7425df8 to 1d0e089 on paper 1.19.2 with SQLite and MariaDB and encountered no issues with it.

Smart123s commented:
@games647 The code is ready for review (and merge?), when you have some spare time. No need to rush.

Smart123s commented:
Hi, I know it's been a long time, but I kind of want to rewrite some parts.

While I remember that we had discussions about the Migrations table, and wether or not it's neccesarry. I've looked at the code of AuthMe, and they don't use a Migrations table. Here's what they use:

https://github.com/AuthMe/AuthMeReloaded/blob/master/src/main/java/fr/xephi/authme/datasource/MySQL.java#L191-L301
https://github.com/AuthMe/AuthMeReloaded/blob/master/src/main/java/fr/xephi/authme/datasource/MySqlMigrater.java

Do you think something similar would be sufficient here?

The commit history could also take some squashing here and there. (and I'll also sign the commits you've pushed).

I saw that there's an [integration-test-containers](https://github.com/games647/FastLogin/tree/integration-test-containers) branch. Either that should be completed, or at least some JUnit tests should be added to make sure that database migrations won't break anytime soon.

games647 commented:
>The commit history could also take some squashing here and there. (and I'll also sign the commits you've pushed).

**Warning I lost some commits during rebase. Maybe you could restore your version or I'll try to dig into my reflog.** 

>Do you think something similar would be sufficient here?

To be honest: This migration concept could be a little too much for a single database change (atm). The idea of migrations comes from my experience with web projects where I saw it quite often with big benefits for multiple contributors (separating logical changes to different files) and being able to rollback. In the end, I'm fine with the AuthMe system too.

>I saw that there's an [integration-test-containers](https://github.com/games647/FastLogin/tree/integration-test-containers) branch.

Yeah the idea was to test core functionality against a full Minecraft server. However, this is very hard to setup and optimize. Nevertheless, it could be very helpful for verifying aspects like login support. 

Regarding the database system, this could be a lot easier and a good idea. By separating the encapsulating the database implementation from the Bukkit API we could be fine with only running a MySQL/MariaDB image.

Smart123s commented:
Let's not forget to check this on Velocity since https://github.com/games647/FastLogin/pull/1071 got merged. This PR only changes code in core, so it's possible, that it'll work without any modifications, but who knows? I'll be pretty busy nowdays, I'll report back once I have some time.

games647 commented on core/src/main/java/com/github/games647/fastlogin/core/shared/FloodgateManagement.java:
```diff
@@ -119,15 +153,12 @@ public void run() {
 
-        if (!isRegistered && !isAutoAuthAllowed(autoRegisterFloodgate)) {
+        if ((!isRegistered && !isAutoAuthAllowed(autoRegisterFloodgate))
+                || (profile.getFloodgate() == FloodgateState.LINKED && !isLinked)
```
Could you please explain the different cases and why they are made. For me it's hard to understand what is actually happening here. Maybe you can extract them to readable methods if that helps.

games647 reviewed (changes requested):
Finally I have some free time so I looked over the code. I would like to merge it now and revive some work on the project. I only inverted some ifs, because for me personally reading the non-inverted-case is easier to read than having shorter code.

In the end, I only have trouble to understand the if-cases in FloodgateManagement that would like to have resolved before merging it. Thank you in advance for the help.

Smart123s commented:
> Finally I have some free time so I looked over the code. I would like to merge it now and revive some work on the project. I only inverted some ifs, because for me personally reading the non-inverted-case is easier to read than having shorter code.
> 
> In the end, I only have trouble to understand the if-cases in FloodgateManagement that would like to have resolved before merging it. Thank you in advance for the help.

Those if statements got really complicated, that's for sure. Some months ago I tried looking into it to simplify the code, but I couldn't really do anything about it.

I have force pushed those changes as a fixup of Differentiate Floodgate players during login & rebased the branch on main, so you may have to re-sign your commits.

It's nice to hear from you after all this time! Thanks for remembering, and looking into my code after so long Hope you are alright!
